File: agent-workflow-app/tools/hardware-controller.js
const readline = require('readline');
const WebSocket = require('ws');

class HardwareController {
  constructor(piUrl = process.env.EMPIRE_WS || 'ws://pi5.local:8765') {
    this.piUrl = piUrl;
    this.ws = null;
    this.pendingApprovals = new Map();
  }

  connect() {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(this.piUrl);
      const timer = setTimeout(() => {
        ws.close();
        reject(new Error('hardware bus connection timed out'));
      }, 3000);

      ws.on('open', () => {
        clearTimeout(timer);
        this.ws = ws;
        this.ws.send(JSON.stringify({
          type: 'register',
          agent_id: 'hardware_controller',
          role: 'engineering'
        }));
        console.log('[HARDWARE] Connected to Pi 5');
        resolve();
      });

      ws.on('error', reject);
    });
  }

  async execute(device, action, requiresApproval = true) {
    if (requiresApproval) {
      const approved = await this.requestApproval(device, action);
      if (!approved) {
        console.log('[HARDWARE] Command cancelled');
        return false;
      }
    }

    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('hardware bus is not connected');
    }

    this.ws.send(JSON.stringify({
      type: 'command',
      target: device,
      payload: { action }
    }));

    return true;
  }

  requestApproval(device, action) {
    return new Promise((resolve) => {
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
      });

      const timer = setTimeout(() => {
        rl.close();
        resolve(false);
      }, 30000);

      rl.question(`[HARDWARE APPROVAL] Execute "${action}" on "${device}"? (y/n): `, (answer) => {
        clearTimeout(timer);
        rl.close();
        resolve(answer.trim().toLowerCase() === 'y');
      });
    });
  }
}

module.exports = HardwareController;

File: tools/hardware_controller.py
#!/usr/bin/env python3
import asyncio
import json
import os
import time

import websockets


class HardwareController:
    def __init__(self, url=None):
        self.url = url or os.getenv("EMPIRE_WS", "ws://pi5.local:8765")
        self.approval_required = os.getenv("HARDWARE_APPROVAL_REQUIRED", "true").lower() != "false"
        self.ws = None

    async def connect(self):
        self.ws = await websockets.connect(self.url)
        await self.ws.send(json.dumps({
            "type": "register",
            "agent_id": "hardware_controller",
            "role": "engineering",
        }))
        print("[HARDWARE] Connected to agent bus")

    async def execute(self, device, action):
        if self.approval_required and not self.request_approval(device, action):
            print("[HARDWARE] Command cancelled")
            return False

        if not self.ws:
            await self.connect()

        await self.ws.send(json.dumps({
            "type": "command",
            "target": device,
            "command": action,
            "approved": True,
            "timestamp": time.time(),
        }))
        return True

    def request_approval(self, device, action):
        answer = input(f"[HARDWARE APPROVAL] Execute '{action}' on '{device}'? (y/n): ")
        return answer.strip().lower() == "y"


async def main():
    controller = HardwareController()
    await controller.connect()
    print("[HARDWARE] Ready. Import this module or call execute() from a supervised command.")


if __name__ == "__main__":
    asyncio.run(main())
